Enables plugin to analyze kotlin projects

rakshitkr opened the pull request:
Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Enables plugin to analyze kotlin projects

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Fix for analyzing eclipse created kotlin-maven-project

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Implemented Eclipse Job for builder invocation

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Added missing KotlinCompiler call for non-maven projects

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Implemented wait functionality for all build & refresh jobs before triggering Kotlin Compiler

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Added kotlin jar to get rid of unresolved imports when kotlin plugin is uninstalled

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Moved kotlin functionality to separate plugin

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
Incorporating review comments

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 					if(ip.hasNature("org.eclipse.m2e.core.maven2Nature")) {
+ 						if(KotlinUtils.verifyKotlinDependency(javaProject)) {
+ 							
+ 							System.out.println("Custom Builder started.");
```
Use Activator.getDefault().logInfo() if you really want to log something.
In this case, be more specific than "custom builder". A user would have no idea, what builder this message is talking about.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 						if(KotlinUtils.verifyKotlinDependency(javaProject)) {
+ 							
+ 							System.out.println("Custom Builder started.");
+ 							final Job builder = new Job("Custom Builder") {
```
Same here.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+									try {
+										ip.build(IncrementalProjectBuilder.CLEAN_BUILD, null);
+									} catch (CoreException e) {
+										e.printStackTrace();
```
All exception logging should go through Activator.getDefault.logError().

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 							while(builder.getState() != Job.NONE) {
+								Thread.sleep(3);
+							}
+ 							System.out.println("Custom Builder finished.");
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 							KotlinUtils.waitForBuildAndRefreshJobs();
+ 							
+							KotlinCompilerResult result = KotlinCompiler.compileKotlinFiles(javaProject);
+							if(result.compiledCorrectly())
```
Add curly braces, even if there is just one line in the branch.

Also, inline "result". The variable is only used once.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+								Activator.getDefault().logInfo("Cannot compile some kotlin files due to errors. Static analysis skipped them.");
+ 						}
+ 						else {
+ 							Activator.getDefault().logInfo("Cannot compile kotlin files without dependency.");
```
Be more specific concerning which dependency you are talking about.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 						}
+ 					}
+ 					else {
+ 						KotlinCompilerResult result = KotlinCompiler.compileKotlinFiles(javaProject);
```
Same comments as above.
If I see this correctly, this branch is the same as the block above. If so, refactor both into one method.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+							Activator.getDefault().logInfo("Cannot compile some kotlin files due to errors. Static analysis skipped them.");
+ 					}	
+ 				} catch (CoreException | OperationCanceledException | InterruptedException e) {
+ 					e.printStackTrace();
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 					e.printStackTrace();
+ 				}
+ 			}
+ 		}
```
Shouldn't there be some kind of handling of cases, when a project does not have a kotlin nature?

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+		IJavaProject javaProject = JavaCore.create(ip);
+		
+		if(Platform.getBundle("org.jetbrains.kotlin.core") != null) {
+ 			if(KotlinNature.hasKotlinNature(ip)) {
```
If the conditions in l48 and l49 need to both be true for anything to happen, just conjunct them in one if clause.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+					srcFilename = Arrays.asList(classFileName.split("\\$")).get(0) + ".kt";
+				}
+				// because in some projects the class names aren't renamed
+				else
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+				String srcFilename = "";
+
+				if(classFileName.substring(classFileName.length()-2).equals("Kt")) {
+					srcFilename = classFileName.substring(0, classFileName.length()-2) + ".kt";
```
Extract .kt as constant.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+						String packageName = String.join(File.separator, modifiedPath);
+
+						IFile sourceFile = currentProject.getFile(packageName + File.separator + srcFilename);
+						if(sourceFile.exists())
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+	}
+	
+	public static boolean verifyKotlinDependency(IJavaProject javaProject) {
+ 		IProject p = javaProject.getProject();
```
Use clearer variable name.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 			for (Dependency dependency : dependencies) {
+ 				System.out.println(dependency);
+ 				if(dependency.getArtifactId().equals("kotlin-stdlib"))
+ 					isPresent = true;
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 			List<Dependency> dependencies = project.getDependencies();
+ 			
+ 			for (Dependency dependency : dependencies) {
+ 				System.out.println(dependency);
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 			
+ 			for (Dependency dependency : dependencies) {
+ 				System.out.println(dependency);
+ 				if(dependency.getArtifactId().equals("kotlin-stdlib"))
```
Change to "kotlin-stdlib".equals(dependency.getArtifactId(). Avoids some potential NPEs.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 					isPresent = true;
+ 			}
+ 			if(!isPresent) {
+ 				boolean accepted = requestUsersPermission();
```
Inline.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 				}
+ 			}
+ 		} catch (Exception e) {
+ 			e.printStackTrace();
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+			jobs.addAll(Arrays.asList(Job.getJobManager().find(ResourcesPlugin.FAMILY_MANUAL_BUILD)));
+			jobs.addAll(Arrays.asList(Job.getJobManager().find(ResourcesPlugin.FAMILY_MANUAL_REFRESH)));
+			if(jobs.isEmpty()) {
+				System.out.println("No pending jobs found.");
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+				System.out.println("No pending jobs found.");
+				return;
+			}
+			System.out.println("Waiting for " + jobs.size() + " Jobs to finish...");
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+	private static void addKotlinDependency(MavenProject project, List<Dependency> dependencies) {
+		Dependency kotlinDependency = new Dependency();
+		kotlinDependency.setGroupId("org.jetbrains.kotlin");
+		kotlinDependency.setArtifactId("kotlin-stdlib");
```
Is there a way to make this more dynamic? We likely always want the latest version of the Kotlin plugin to be added, not always Kotline 1.3.61, right?

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+ 			mavenWriter.write(writer, model);
+ 			Activator.getDefault().logInfo("Required kotlin dependency added to pom.");
+ 		} catch (IOException e) {
+ 			e.printStackTrace();
```
.

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+
+ 		if (pomFile != null && pomFile.exists())
+ 		{
+ 			FileReader reader = null;
```
Rather use try with resource: https://docs.oracle.com/en/java/javase/13/language/try-resources.html

kruegers commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,261 @@
+		dialog.setMessage("Analysis requires maven kotlin-stdlib dependency. Would you like to add it?");
+ 		dialog.setText("CAUTION");
+ 		int opt = dialog.open();
+ 		if (opt == SWT.YES)
```
Change to "return dialog.open();"

kruegers commented on plugins/de.cognicrypt.staticanalyzer/.classpath:
```diff
@@ -1,7 +1,7 @@
-	<classpathentry kind="src" path="src"/>
-	<classpathentry kind="output" path="target/classes"/>
-</classpath>
+<?xml version="1.0" encoding="UTF-8"?>
```
This file should stay as it is.

kruegers commented on plugins/de.cognicrypt.staticanalyzer/META-INF/MANIFEST.MF:
```diff
@@ -23,4 +23,5 @@ Bundle-ActivationPolicy: lazy
  resources/
 Automatic-Module-Name: de.cognicrypt.staticanalyzer
-Import-Package: org.eclipse.ui.texteditor
+Import-Package: de.cognicrypt.staticanalyzer.kotlin.utilities,
```
If there is a dependency between these two plugins, it should be the other way around. Please check why this dependency exists and resolve it differently.

kruegers commented on plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/handlers/AnalysisKickOff.java:
```diff
@@ -89,9 +91,12 @@ public boolean setUp(final IJavaElement iJavaElement) {
-		return true;
-	}
+		
+		KotlinUtils.compileKotlinFiles(ip);
```
Do not call the kotlin plugin from inside this one like this. This call causes the kotlin plugin to be required under all circumstances.

rakshitkr replied:
Could you please suggest me some way to achieve this? I've implemented rest of your review changes except this one.

krinara86 replied:
I think reflection or some kind of dependency injection would be the way to go. Reflection is not that clean.  We should look up some dependency injection frameworks that are clean. This sounds tricky to accomplish but a needed requirement as we do not want to have dependency from kotlin to base.

krinara86 replied:
One option is to use the factory pattern. Have a base factory runner that calls the kotlin runner if configuration is kotlin. https://www.tutorialspoint.com/design_pattern/factory_pattern.htm

rakshitkr committed:
Restoring changes before commit f2e87d93

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

rakshitkr committed:
This commit incorporates review changes from PR-#378

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

Below changes have been made:
- replaced sysouts with logInfos
- changed exception logging to go through Activator.getDefault.logError()
- added curly braces to single line branch code
- removed single use variables
- used try-with-resource syntax instead of try-finally
- changed return type of selectedDialogOption method

rakshitkr committed:
Removed kotlin dependency from staticanalyzer plugin

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

- implemented extension
- staticanalyzer plugin is the host plugin
- kotlin plugin is the extender plugin that gets triggered when the project is of kotlin nature
- this would would make staticanalyzer plugin independent of kotlin plugin

rakshitkr committed:
Merge branch 'develop' into kotlin-fix-v3

rakshitkr committed:
Replaced kotlin.Activator with staticanalyzer.Activator

Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

- this is done because there is no Activator instance for kotlin plugin since its invoked as a listener by staticanalyzer

rakshitkr committed:
Merge branch 'develop' into kotlin-fix-v3

rakshitkr committed:
Merge remote-tracking branch 'upstream/develop' into kotlin-fix-v3

rakshitkr commented:
The sample projects for testing kotlin plugin can be found [here](https://github.com/CROSSINGTUD/CryptoAnalysis/commit/bae23f3d83ea99ea9668a06bb84274a33e9556bb)

AnakinRaW commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,247 @@
+ 							
+ 							while(builder.getState() != Job.NONE) {
+								Thread.sleep(3);
+							}
```
Do we need to sleep or could we just Thread.yield()?

AnakinRaW commented on plugins/de.cognicrypt.staticanalyzer.kotlin/src/de/cognicrypt/staticanalyzer/kotlin/utilities/KotlinUtils.java:
```diff
@@ -0,0 +1,247 @@
+			}
+		}
+		
+		throw new ClassNotFoundException("Class " + className + " not found.");
```
Either use a CogniCryptException or IOException

AnakinRaW commented on plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/IListener.java:
```diff
@@ -0,0 +1,10 @@
+import org.eclipse.jdt.core.JavaModelException;
+
+public interface IListener {
+	void listen1(IProject ip);
```
Do we have better names?

AnakinRaW commented on plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/utilities/Utils.java:
```diff
@@ -0,0 +1,53 @@
+ 	        }
+ 		}
+ 		catch (final JavaModelException e) {
+ 			throw new ClassNotFoundException("Class " + className + " not found.", e);
```
same as above
